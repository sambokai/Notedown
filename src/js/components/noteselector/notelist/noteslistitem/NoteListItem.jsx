import React from 'react';
import PropTypes from 'prop-types';
import Moment from 'moment/moment';


class NoteListItem extends React.Component {
  static relativeCalendarFormat = {
    sameDay: 'HH:mm',
    lastDay: '[Yesterday]',
    sameElse: 'DD.MM.YY',
  };

  static getRelativeCalendarDate(date) {
    return Moment(date).calendar(null, this.relativeCalendarFormat);
  }

  getNoteTitle() {
    if (this.props.note.body) {
      return this.props.note.body.match(/(\n|\s)*.*(\n|$)/)[0].trim();
    }

    return this.props.emptyNoteTitle;
  }

  render() {
    return (
      <a
        ref={this.props.draggableProvided.innerRef}
        {...this.props.draggableProvided.draggableProps}
        {...this.props.draggableProvided.dragHandleProps}
        role="button"
        tabIndex={-1}
        className={
          `flex-column align-items-center list-group-item list-group-item-action px-2
              ${(this.props.isSelected) ? 'active' : ''}
              ${this.props.note.body ? '' : 'disabled font-weight-light bg-secondary text-white'}`
        }
        onClick={() => this.props.onSelectNote(this.props.note.id)}
        onKeyDown={() => this.props.onSelectNote(this.props.note.id)}
      >
        <div className="d-flex w-100 align-items-center justify-content-between">
          <p id="note-list-item-title" className="text-truncate pr-2 mb-0">{this.getNoteTitle()}</p>
          <small >
            {NoteListItem.getRelativeCalendarDate(this.props.note.lastChange)}
          </small>
        </div>
      </a>
    );
  }
}

NoteListItem.propTypes = {
  note: PropTypes.shape({
    body: PropTypes.string,
    id: PropTypes.number,
    lastChange: PropTypes.number,
  }).isRequired,
  onSelectNote: PropTypes.func.isRequired,
  emptyNoteTitle: PropTypes.string,
  isSelected: PropTypes.bool.isRequired,

  draggableProvided: PropTypes.shape({
    innerRef: PropTypes.func,
    draggableProps: PropTypes.object,
    dragHandleProps: PropTypes.object,
  }),
};


NoteListItem.defaultProps = {
  emptyNoteTitle: 'Empty',
  draggableProvided: {},
};

export default NoteListItem;
