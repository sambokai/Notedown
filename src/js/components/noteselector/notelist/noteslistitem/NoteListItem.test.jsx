/* eslint-disable comma-dangle,function-paren-newline */

import React from 'react';

import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import NoteListItem from './NoteListItem';

let dateNowSpy;
const lockedTime = new Date(2010, 6, 15, 12).valueOf();

beforeAll(() => {
  // Lock Time
  dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => lockedTime);
});

afterAll(() => {
  // Unlock Time
  dateNowSpy.mockReset();
  dateNowSpy.mockRestore();
});

const onSelectNoteMock = jest.fn();
onSelectNoteMock.mockReturnValue('This Note was selected');

const mockProps = {
  note: {
    body: 'This is Note #1',
    id: 99,
    lastChange: new Date(Date.UTC(2018, 0, 1)).valueOf()
  },
  draggableProvided: {},
  onSelectNote: onSelectNoteMock
};

describe('<NoteListItem/>', () => {
  it('renders correctly if the item is selected', () => {
    const wrapper = renderer
      .create(
        <NoteListItem {...mockProps} isSelected />
      );

    expect(wrapper)
      .toMatchSnapshot();
  });

  it('renders correctly if the item is NOT selected', () => {
    const wrapper = renderer
      .create(
        <NoteListItem {...mockProps} isSelected={false} />
      );

    expect(wrapper)
      .toMatchSnapshot();
  });

  it("calls onSelectNote callback, upon click with the note's id as the function argument", () => {
    const wrapper = shallow(<NoteListItem {...mockProps} isSelected={false} />);

    wrapper.find('a').simulate('click');

    expect(onSelectNoteMock.mock.calls).toHaveLength(1);
    expect(onSelectNoteMock.mock.calls[0][0]).toBe(mockProps.note.id);
  });

  it("calls onSelectNote callback, upon keyDown with the note's id as the function argument", () => {
    const wrapper = shallow(<NoteListItem {...mockProps} isSelected={false} />);

    wrapper.find('a').simulate('keyDown');

    expect(onSelectNoteMock.mock.calls).toHaveLength(1);
    expect(onSelectNoteMock.mock.calls[0][0]).toBe(mockProps.note.id);
  });
});

describe('getNoteTitle()', () => {
  it("returns the first line of text in a note's body as a trimmed string", () => {
    const testNote = {
      ...mockProps.note,
      body: ' \n \n    This is the first line of Note #1     \n \n ' +
      'Yesterday I went to the mall. It was great.'
    };

    const wrapper = shallow(<NoteListItem isSelected note={testNote} onSelectNote={mockProps.onSelectNote} />);

    const result = wrapper.instance().getNoteTitle();

    expect(result).toBe('This is the first line of Note #1');
  });

  it("returns a predefined 'Empty'-Message if the note's body is empty", () => {
    const testNote = {
      ...mockProps.note,
      body: ''
    };
    const emptyNoteTitle = 'Empty';

    const wrapper = shallow(
      <NoteListItem isSelected note={testNote} onSelectNote={mockProps.onSelectNote} emptyNoteTitle={emptyNoteTitle} />
    );

    const result = wrapper.instance().getNoteTitle();

    expect(result).toBe(emptyNoteTitle);
  });
});

describe('getRelativeCalendarDate(date)', () => {
  it('returns exact time, if the provided date was today', () => {
    const lastChange = new Date(lockedTime);
    lastChange.setHours(lastChange.getHours() - 6);

    const relativeDate = NoteListItem.getRelativeCalendarDate(lastChange);
    expect(relativeDate).toBe('06:00');
  });

  it('returns \'Yesterday\', if the provided date was yesterday', () => {
    const lastChange = new Date(lockedTime);
    lastChange.setHours(lastChange.getHours() - 24);

    const relativeDate = NoteListItem.getRelativeCalendarDate(lastChange);
    expect(relativeDate).toBe('Yesterday');
  });

  it('returns exact date, if the provided date was more than 24 hours ago', () => {
    const lastChange = new Date(lockedTime);
    lastChange.setHours(lastChange.getHours() - (24 * 90));

    const relativeDate = NoteListItem.getRelativeCalendarDate(lastChange);
    expect(relativeDate).toBe('16.04.10');
  });

  it('returns exact date, if the provided date was more than 24 hours, but less than a week ago', () => {
    const lastChange = new Date(lockedTime);
    lastChange.setHours(lastChange.getHours() - (24 * 5));

    const relativeDate = NoteListItem.getRelativeCalendarDate(lastChange);
    expect(relativeDate).toBe('10.07.10');
  });
});
