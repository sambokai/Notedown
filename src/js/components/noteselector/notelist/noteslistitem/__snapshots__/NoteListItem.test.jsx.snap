// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteListItem/> renders correctly 1`] = `
<a
  className="flex-column align-items-center list-group-item list-group-item-action px-2
              active
              "
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={-1}
>
  <div
    className="d-flex w-100 align-items-center justify-content-between"
  >
    <p
      className="text-truncate pr-2 mb-0"
      id="note-list-item-title"
    >
      This is Note #1
    </p>
    <small>
      01.01.18
    </small>
  </div>
</a>
`;

exports[`renders correctly 1`] = `
<a
  className="flex-column align-items-center list-group-item list-group-item-action px-2
              active
              "
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={-1}
>
  <div
    className="d-flex w-100 align-items-center justify-content-between"
  >
    <p
      className="text-truncate pr-2 mb-0"
      id="note-list-item-title"
    >
      This is Note #1
    </p>
    <small>
      01.01.18
    </small>
  </div>
</a>
`;
